From 11bc7a256c6c37b70af33451f0e651165e7d9dd4 Mon Sep 17 00:00:00 2001
From: renchenghan <renchenghan@vivo.corp-partner.google.com>
Date: Fri, 22 Aug 2025 16:21:51 +0800
Subject: [PATCH 08/15] ANDROID: vendor_hooks: add hook in
 __dev_pm_qos_update_request()

This hook is used to update new values to req->data.freq. When the GPU
updates its maximum frequency by __dev_pm_qos_update_request(), this
hook allows us to capture the frequency information, enabling
centralized management of requests from different sources

Bug: 469940255
Bug: 440488476
Signed-off-by: renchenghan <renchenghan@vivo.corp-partner.google.com>
Change-Id: Ieadb004f659dd9e7db0b6632b268fc7b440292a4
(cherry picked from commit 4373bc5d8ad18e4d51c1efc114c13bfb123cd4fa)
---
 drivers/android/vendor_hooks.c | 1 +
 drivers/base/power/qos.c       | 7 +++++--
 include/trace/hooks/power.h    | 4 ++++
 3 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/drivers/android/vendor_hooks.c b/drivers/android/vendor_hooks.c
index f41f1311cb4d..93bbe26d2947 100644
--- a/drivers/android/vendor_hooks.c
+++ b/drivers/android/vendor_hooks.c
@@ -644,6 +644,7 @@ EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_ksize);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_kmalloc_large_node_bypass);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_kfree_bypass);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_check_heap_object_bypass);
+EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_fas_gpu_qos_update_tracer);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_printk_save_irq);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_handle_add_fec_mismatch_blks);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_handle_data_error);
diff --git a/drivers/base/power/qos.c b/drivers/base/power/qos.c
index ff393cba7649..b818463593b3 100644
--- a/drivers/base/power/qos.c
+++ b/drivers/base/power/qos.c
@@ -36,6 +36,7 @@
 #include <linux/pm_runtime.h>
 #include <linux/err.h>
 #include <trace/events/power.h>
+#include <trace/hooks/power.h>
 
 #include "power.h"
 
@@ -437,9 +438,11 @@ static int __dev_pm_qos_update_request(struct dev_pm_qos_request *req,
 
 	trace_dev_pm_qos_update_request(dev_name(req->dev), req->type,
 					new_value);
-	if (curr_value != new_value)
+	if (curr_value != new_value) {
+		if (req->type == DEV_PM_QOS_MAX_FREQUENCY)
+			trace_android_vh_fas_gpu_qos_update_tracer(&req->data.freq, &new_value);
 		ret = apply_constraint(req, PM_QOS_UPDATE_REQ, new_value);
-
+	}
 	return ret;
 }
 
diff --git a/include/trace/hooks/power.h b/include/trace/hooks/power.h
index 5721b3ddf611..f4544ce9acba 100644
--- a/include/trace/hooks/power.h
+++ b/include/trace/hooks/power.h
@@ -45,6 +45,10 @@ DECLARE_HOOK(android_vh_append_total_power,
 			u64 total_stime, u64 total_power),
 		TP_ARGS(m, uid, total_utime, total_stime, total_power));
 
+DECLARE_HOOK(android_vh_fas_gpu_qos_update_tracer,
+		TP_PROTO(struct freq_qos_request *req, s32 *value),
+		TP_ARGS(req, value));
+
 #endif /* _TRACE_HOOK_POWER_H */
 /* This part must be outside protection */
 #include <trace/define_trace.h>
-- 
2.50.1

